### PR TITLE
Implement packing local variable and constant data on 64-bit if stack usage is high

### PR DESCRIPTION
**Pull request by ksh8281**

### CODE REVIEW

File: src/parser/WASMParser.cpp [hunk 1]
```diff
@@ -1007,29 +1011,90 @@ class WASMBinaryReader : public wabt::WASMBinaryReaderDelegate {
...
+#define WALRUS_ENABLE_LOCAL_VARIABLE_PACKING_MIN_SIZE 64
+#endif
+        // pack local variables if needs
+        constexpr size_t enableLocalVaraiblePackingMinSize = WALRUS_ENABLE_LOCAL_VARIABLE_PACKING_MIN_SIZE;
+        if (m_initialFunctionStackSize >= enableLocalVaraiblePackingMinSize) {
+            m_initialFunctionStackSize = m_currentFunctionType->paramStackSize();
+            // put already aligned variables first
+            for (size_t i = m_currentFunctionType->param().size(); i < m_localInfo.size(); i++) {
+                auto& info = m_localInfo[i];
+                if (Walrus::hasCPUWordAlingedSize(info.m_valueType) || needsCPUWordAlingedAddress(info.m_valueType)) {
```
Comment (clover2123) [inline src/parser/WASMParser.cpp:1033]: a typo `Alinged` -> `Aligned`

File: src/runtime/Value.h [hunk 2]
```diff
@@ -431,10 +431,34 @@ inline size_t valueFunctionCopyCount(Value::Type type)
...
+inline bool needsCPUWordAlingedAddress(Value::Type type)
+{
+#if defined(WALRUS_32)
+    // everything is already aligned!
+    return false;
+#else
+    switch (type) {
+    case Value::FuncRef:
+    case Value::ExternRef:
+        return true;
```
Comment (clover2123) [inline src/runtime/Value.h:451]: little bit confused, do these FuncRef and ExternRef need alignment?

File: src/parser/WASMParser.cpp (hunk 1 above)
Comment (ksh8281) [reply src/parser/WASMParser.cpp:1033]: Oops!

File: src/Walrus.h [hunk 3]
```diff
@@ -347,7 +347,7 @@ if (f.type == Type::B) { puts("failed in msvc."); }
     std::unique_ptr<uint8_t[]> Result##HolderWhenUsingMalloc;                                   \
     size_t bytes##Result = (Bytes);                                                             \
     Type* Result;                                                                               \
-    if (LIKELY(bytes##Result < 512)) {                                                          \
+    if (LIKELY(bytes##Result < 2048)) {                                                         \
```
[the remark this answers is not in the document]
Comment (ksh8281) [reply src/Walrus.h:350]: Code generated by emcc sometimes uses stack a lot.

File: src/runtime/Value.h (hunk 2 above)
Comment (ksh8281) [reply src/runtime/Value.h:451]: Arm devices cannot support unaligned pointer well :(
https://embeddedartistry.com/blog/2016/12/26/arm-pointer-alignment-requirements/
+ if we want to use bdwgc later, we need to align pointer in stack

Comment (clover2123) [reply src/runtime/Value.h:451]: Then what about 32bit platform?
This patch marks alignment only for addresses in 64bit platform, but I think that this alignment needs to be applied for 32bit addresses too

Comment (ksh8281) [reply src/runtime/Value.h:451]: IMO we don't consider alignment of 32-bit platform
Every value contents have multiple of 4-byte size and
and we have experience that alloca function returns aligned address

Comment (clover2123) [reply src/runtime/Value.h:451]: Oh, I get it :)
Packing approach has this issue only for 64bit platform

**Review by clover2123 — APPROVED** [review]
LGTM

**Pull request merged**